File: scripts/fetch_youtube_comments.py
#!/usr/bin/env python3
"""
Fetches recent YouTube comments for a channel and outputs structured JSON.
Usage: YOUTUBE_API_KEY=... YOUTUBE_CHANNEL_ID=... python3 fetch_youtube_comments.py
"""

import json
import os
import sys
import requests
from datetime import datetime, timedelta, timezone

YOUTUBE_API_KEY = os.environ.get("YOUTUBE_API_KEY", "")
YOUTUBE_CHANNEL_ID = os.environ.get("YOUTUBE_CHANNEL_ID", "")
DAYS_BACK = int(os.environ.get("DAYS_BACK", "7"))
MAX_VIDEOS = int(os.environ.get("MAX_VIDEOS", "10"))
MAX_COMMENTS_PER_VIDEO = int(os.environ.get("MAX_COMMENTS_PER_VIDEO", "100"))
BASE_URL = "https://www.googleapis.com/youtube/v3"


def api_get(endpoint, params):
    params["key"] = YOUTUBE_API_KEY
    resp = requests.get(f"{BASE_URL}/{endpoint}", params=params, timeout=15)
    if not resp.ok:
        error = resp.json().get("error", {})
        raise RuntimeError(f"YouTube API error on /{endpoint}: {error.get('message', resp.text)}")
    return resp.json()


def get_recent_videos(channel_id):
    since = (datetime.now(timezone.utc) - timedelta(days=DAYS_BACK)).strftime("%Y-%m-%dT%H:%M:%SZ")
    data = api_get("search", {
        "part": "id,snippet",
        "channelId": channel_id,
        "type": "video",
        "order": "date",
        "publishedAfter": since,
        "maxResults": MAX_VIDEOS,
    })
    videos = [
        {
            "id": item["id"]["videoId"],
            "title": item["snippet"]["title"],
            "published": item["snippet"]["publishedAt"],
            "description": item["snippet"]["description"][:200],
        }
        for item in data.get("items", [])
        if item.get("id", {}).get("videoId")
    ]
    if not videos:
        data = api_get("search", {
            "part": "id,snippet",
            "channelId": channel_id,
            "type": "video",
            "order": "date",
            "maxResults": MAX_VIDEOS,
        })
        videos = [
            {
                "id": item["id"]["videoId"],
                "title": item["snippet"]["title"],
                "published": item["snippet"]["publishedAt"],
                "description": item["snippet"]["description"][:200],
            }
            for item in data.get("items", [])
            if item.get("id", {}).get("videoId")
        ]
    return videos


def get_comments(video_id):
    try:
        data = api_get("commentThreads", {
            "part": "snippet",
            "videoId": video_id,
            "order": "relevance",
            "maxResults": MAX_COMMENTS_PER_VIDEO,
            "textFormat": "plainText",
        })
    except RuntimeError as e:
        return [], str(e)

    comments = []
    for item in data.get("items", []):
        top = item["snippet"]["topLevelComment"]["snippet"]
        comments.append({
            "author": top.get("authorDisplayName", ""),
            "text": top.get("textDisplay", ""),
            "likes": top.get("likeCount", 0),
            "published": top.get("publishedAt", ""),
        })
    return comments, None


def main():
    errors = []

    if not YOUTUBE_API_KEY:
        errors.append("YOUTUBE_API_KEY environment variable not set")
    if not YOUTUBE_CHANNEL_ID:
        errors.append(
            "YOUTUBE_CHANNEL_ID environment variable not set. "
            "Find your channel ID at youtube.com → Your channel → About → Share → Copy channel ID"
        )

    if errors:
        print(json.dumps({"errors": errors}))
        sys.exit(1)

    try:
        videos = get_recent_videos(YOUTUBE_CHANNEL_ID)
    except RuntimeError as e:
        print(json.dumps({"errors": [str(e)]}))
        sys.exit(1)

    results = []
    for video in videos:
        comments, err = get_comments(video["id"])
        entry = {
            "video_id": video["id"],
            "video_title": video["title"],
            "published": video["published"],
            "url": f"https://www.youtube.com/watch?v={video['id']}",
            "comment_count": len(comments),
            "comments": comments,
        }
        if err:
            entry["error"] = err
        results.append(entry)

    total = sum(v["comment_count"] for v in results)
    print(json.dumps({
        "channel_id": YOUTUBE_CHANNEL_ID,
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "days_back": DAYS_BACK,
        "videos_found": len(results),
        "total_comments": total,
        "videos": results,
    }, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: scripts/fetch-youtube-comments.js
#!/usr/bin/env node
// Fetches recent video comments from a YouTube channel using the Data API v3.
// Outputs JSON to stdout for analysis by Claude.
//
// Required env vars (set in scripts/.env or export manually):
//   YOUTUBE_API_KEY   - YouTube Data API v3 key
//   YOUTUBE_CHANNEL_ID - Your channel ID (e.g. UCxxxxxxxxxxxxxxx)
//
// Optional:
//   MAX_VIDEOS   - how many recent videos to scan (default: 5)
//   MAX_COMMENTS - comments per video (default: 50)

const https = require('https');

const API_KEY     = process.env.YOUTUBE_API_KEY;
const CHANNEL_ID  = process.env.YOUTUBE_CHANNEL_ID;
const MAX_VIDEOS  = parseInt(process.env.MAX_VIDEOS  || '5',  10);
const MAX_COMMENTS= parseInt(process.env.MAX_COMMENTS|| '50', 10);

if (!API_KEY || !CHANNEL_ID) {
  console.error(JSON.stringify({
    error: 'Missing configuration',
    missing: [
      ...(!API_KEY     ? ['YOUTUBE_API_KEY']    : []),
      ...(!CHANNEL_ID  ? ['YOUTUBE_CHANNEL_ID'] : []),
    ],
    hint: 'Set these in scripts/.env or as environment variables.',
  }));
  process.exit(1);
}

function get(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (parsed.error) reject(new Error(`YouTube API: ${parsed.error.message}`));
          else resolve(parsed);
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

function qs(params) {
  return Object.entries(params)
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
    .join('&');
}

async function getRecentVideos() {
  const params = qs({
    part: 'snippet',
    channelId: CHANNEL_ID,
    maxResults: MAX_VIDEOS,
    order: 'date',
    type: 'video',
    key: API_KEY,
  });
  const data = await get(`https://www.googleapis.com/youtube/v3/search?${params}`);
  return (data.items || []).map(item => ({
    id: item.id.videoId,
    title: item.snippet.title,
    publishedAt: item.snippet.publishedAt,
  }));
}

async function getComments(videoId, videoTitle) {
  const params = qs({
    part: 'snippet',
    videoId,
    maxResults: MAX_COMMENTS,
    order: 'relevance',
    key: API_KEY,
  });
  let data;
  try {
    data = await get(`https://www.googleapis.com/youtube/v3/commentThreads?${params}`);
  } catch (e) {
    // Comments disabled or video not found — skip gracefully
    return { videoId, videoTitle, commentsDisabled: true, comments: [] };
  }

  const comments = (data.items || []).map(item => {
    const top = item.snippet.topLevelComment.snippet;
    return {
      author: top.authorDisplayName,
      text: top.textOriginal,
      likes: top.likeCount,
      publishedAt: top.publishedAt,
    };
  });

  return { videoId, videoTitle, commentsDisabled: false, comments };
}

async function main() {
  const videos = await getRecentVideos();

  if (!videos.length) {
    console.log(JSON.stringify({ channel: CHANNEL_ID, videos: [], message: 'No recent videos found.' }));
    return;
  }

  const results = await Promise.all(videos.map(v => getComments(v.id, v.title)));
  const output = {
    channel: CHANNEL_ID,
    fetchedAt: new Date().toISOString(),
    videoCount: videos.length,
    videos: results,
  };

  console.log(JSON.stringify(output, null, 2));
}

main().catch(err => {
  console.error(JSON.stringify({ error: err.message }));
  process.exit(1);
});
